euphonic/spectra.py

```diff
@@ -669,10 +674,387 @@ def broaden(self: T, x_width,
+
+        for one_line_data in line_data:
+            yield common_metadata | one_line_data
+
+    def _select_indices(self, **select_key_values) -> list[int]:
+        required_metadata = select_key_values.items()
+        indices = [i for i, row in enumerate(self.iter_metadata())
+                   if required_metadata <= row.items()]
+        return indices
+
+    def select(self, **select_key_values: Union[
+            str, int, Sequence[str], Sequence[int]]) -> Self:
+        """
+        Select spectra by their keys and values in metadata['line_data']
+
+        Parameters
+        ----------
+        **select_key_values
+            Key-value/values pairs in metadata['line_data'] describing
+            which spectra to extract. For example, to select all spectra
+            where metadata['line_data']['species'] = 'Na' or 'Cl' use
+            spectrum.select(species=['Na', 'Cl']). To select 'Na' and
+            'Cl' spectra where weighting is also coherent, use
+            spectrum.select(species=['Na', 'Cl'], weighting='coherent')
+
+        Returns
+        -------
+        selected_spectra
+           A Spectrum1DCollection containing the selected spectra
+
+        Raises
+        ------
+        ValueError
+            If no matching spectra are found
+        """
+        # Convert all items to sequences of possibilities
+        select_key_values = dict(
+            (key, (value,)) if isinstance(value, (int, str)) else (key, value)
+            for key, value in select_key_values.items()
+        )
+
+        # Collect indices that match each combination of values
+        selected_indices = []
+        for value_combination in itertools.product(*select_key_values.values()
+                                                   ):
+            selection = dict(zip(select_key_values.keys(), value_combination))
+            selected_indices.extend(self._select_indices(**selection))
+
+        if not selected_indices:
+            raise ValueError(f'No spectra found with matching metadata '
+                             f'for {select_key_values}')
+
+        return self[selected_indices]
+
+    @staticmethod
+    def _combine_metadata(all_metadata: LineData) -> Metadata:
+        """
+        From a sequence of metadata dictionaries, combines all common
+        key/value pairs into the top level of a metadata dictionary,
+        all unmatching key/value pairs are put into the 'line_data'
+        key, which is a list of metadata dicts for each element in
+        all_metadata
+        """
+        # This is for combining multiple separate spectrum metadata,
+        # they shouldn't have line_data
+        for metadata in all_metadata:
+            assert 'line_data' not in metadata.keys()
+
+        # Combine all common key/value pairs into new dict
+        combined_metadata = dict(
+            set(all_metadata[0].items()).intersection(
+                *[metadata.items() for metadata in all_metadata[1:]]))
```

Comment:
I think it can be made a bit more eloquent with
```python
combined_metadata = dict(
    reduce(set.intersection,
           (set(metadata.items()) for metadata in all_metadata)))
```

Is that readable to you?

Reply:
That makes more sense to me, yeah. Not sure whether it will to others though.